sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCastTest.java

```diff
@@ -0,0 +1,137 @@
+                    SerializableFunctions.identity()));
+
+    Schema resultType =
+        Schema.builder().addInt32Field("f_int").addNullableField("f_date", DATETIME).build();
+
+    PCollection<Row> result =
+        input.apply(
+            SqlTransform.query(
+                "SELECT f_int, \n"
+                    + "   CAST( \n"
+                    + "     SUBSTRING(TRIM(f_string) FROM 1 FOR 4) \n"
+                    + "      ||'-' \n"
+                    + "      ||SUBSTRING(TRIM(f_string) FROM 5 FOR 2) \n"
+                    + "      ||'-' \n"
+                    + "      ||SUBSTRING(TRIM(f_string) FROM 7 FOR 2) as DATE) \n"
+                    + "FROM PCOLLECTION"));
+
+    PAssert.that(result)
+        .containsInAnyOrder(
+            Row.withSchema(resultType).addValues(1, new DateTime(2018, 10, 18, 0, 0)).build());
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testCastToDate2() {
+    PCollection<Row> input =
+        pipeline.apply(
+            Create.of(Row.withSchema(INPUT_ROW_SCHEMA).addValues(1).addValue("20181018").build())
+                  .withSchema(
+                      INPUT_ROW_SCHEMA,
+                      SerializableFunctions.identity(),
+                      SerializableFunctions.identity()));
+
+    Schema resultType =
+        Schema.builder().addInt32Field("f_int").addNullableField("f_date", DATETIME).build();
+
+    PCollection<Row> result =
+        input.apply(
+            SqlTransform.query(
+                "SELECT f_int, \n"
+                + "   CAST( \n"
+                + "     f_string AS DATE) \n"
+                + "FROM PCOLLECTION"));
+
+    PAssert.that(result)
+           .containsInAnyOrder(
+               Row.withSchema(resultType).addValues(1, new DateTime(2018, 10, 18, 0, 0)).build());
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testCastToDateWithCase() {
+    PCollection<Row> input =
+        pipeline.apply(
+            Create.of(Row.withSchema(INPUT_ROW_SCHEMA).addValues(1).addValue("20181018").build())
+                .withSchema(
+                    INPUT_ROW_SCHEMA,
+                    SerializableFunctions.identity(),
+                    SerializableFunctions.identity()));
+
+    Schema resultType = Schema.builder().addInt32Field("f_int").addDateTimeField("f_date").build();
+
+    PCollection<Row> result =
+        input.apply(
+            "sqlQuery",
+            SqlTransform.query(
+                "SELECT f_int, \n"
+                    + "CASE WHEN CHAR_LENGTH(TRIM(f_string)) = 8 \n"
+                    + "    THEN CAST (\n"
+                    + "       SUBSTRING(TRIM(f_string) FROM 1 FOR 4) \n"
+                    + "        ||'-' \n"
+                    + "        ||SUBSTRING(TRIM(f_string) FROM 5 FOR 2) \n"
+                    + "        ||'-' \n"
+                    + "        ||SUBSTRING(TRIM(f_string) FROM 7 FOR 2) AS DATE)\n"
+                    + "    ELSE DATE '2001-01-01'\n"
```

Comment:
`ELSE NULL` makes more sense. But since it's time consuming to change Calcite code, this workaround LGTM